llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+			if (const llvm::BasicBlock* from = getIncomingBB())
+			{
+				const llvm::Value* incomingVal = phi->getIncomingValueForBlock(from);
+				assert(incomingVal);
+				if (isValueComputed(incomingVal))
+				{
+					incomingsVal.push_back({phi, incomingVal});
+					return;
+				}
+			}
+			notComputedPhis.push_back(phi);
+			return;	
+		}
+		else
+		{
+			//While visiting the first NonPhi instruction we set the maps state correctly
+			for (auto& p : incomingsVal)
+				assignToMaps(p.first, p.second);
+			
+			for (auto& p : notComputedPhis)
+				   removeFromMaps(p);	
+			
+			//Then we clear the vectors since this has to be done only once
+			incomingsVal.clear();
+			notComputedPhis.clear();
+		}
+
+		//Here PHI have been properly processed
+
+		bool skip = hasToBeSkipped(I);
+
+		if (isBottomStack())
+		{
+			if (I.isTerminator())
+				skip = true;
+
+			//Skip Instructions we don't have enough information to execute
+			if (skip)
+				return;
+		}
+		else
+		{
+			//Terminators have special handling done here explicitly
+			if (I.isTerminator())
+			{
+				const BasicBlock* next = nullptr;
+
+				switch (I.getOpcode())
+				{
+					case Instruction::Br:
+					{
+						const BranchInst& BI = cast<BranchInst>(I);
+						
+						if (BI.isConditional())
+						{
+							if (isValueComputed(BI.getCondition()))
+							{
+								GenericValue V = getOperandValue(BI.getCondition());
+								next = BI.getSuccessor((V.IntVal == 0u) ? 1 : 0);
+							}
+						}
+						else
+							next = BI.getSuccessor(0);
+						break;
+					}
+					case Instruction::Switch:
+					{
+						const SwitchInst& SI = cast<SwitchInst>(I);
+				
+						if (isValueComputed(SI.getCondition()))
+						{
+							GenericValue V = getOperandValue(SI.getCondition());
+							ConstantInt* CI = (ConstantInt*)ConstantInt::get((IntegerType*)(SI.getCondition()->getType()), V.IntVal);
+							next = SI.findCaseValue(CI)->getCaseSuccessor();
+						}
+						break;
+					}
+					case Instruction::Ret:
+					{
+						const ReturnInst& RI = cast<ReturnInst>(I);
+
+						if (RI.getReturnValue())
+							stronglyKnownBits[getFirstStack().Caller] = getBitMask(RI.getReturnValue());
+
+						//Dispatch to the Interpreter's visitor for the given Instructon
+						Interpreter::visit(const_cast<Instruction&>(I));
+
+						//Note that here we return
+						return;
+					}
+					default:
+					{
+						//Default handling of not doing anything is conservative
+						break;
+					}
+				}
+
+				if (next)
+				{
+					//We know where execution should proceed
+					setIncomingBB(I.getParent());
+					getLastStack().CurBB = const_cast<BasicBlock*>(next);
+					getLastStack().CurInst = getLastStack().CurBB->begin();
+					//Also here we have set the proper state for the execution so we can return
+					return;
+				}
+				else
+				{
+					skip = true;
+				}
+			}
+
+			//We are inside a call, here we assume all failure to execute are non-recoverable (as in no information could be gained)
+			if (skip)
```

Comment:
It should be possible to move this check above, and only reason over the various terminators if 'skip' is false. I believe that if skip is false then it is guaranteed that the terminators can be resolved and 'next' will be known.

Reply:
Reworked, should be clearer now (factored out the logic for finding the next BB)